Caching Boundaries using Connection Groups (#294)
ietf-tapswg/api-drafts: draft-ietf-taps-arch.md

The Transport System Implementation Concepts define the set of objects used internally to a system or library to implement the functionality needed to provide a transport service across a network, as required by the abstract interface.

* Connection Group: A set of Connections that share properties. For multiplexing transport protocols, the Connection Group defines the set of Connections that can be multiplexed together. Groups can be defined implicitly by the Implementation, based on which properties are compatible. However, applications SHOULD be able to explicitly define Connection Groups, as discussed in {{groups}}.

> 

Reviewer: Defining the Connection Groups implicitly is a bit different from what we have in the API draft at the moment where I would read it as Connection Groups are formed explicitly by calling Clone, So the API text should probbaly be aligned a bit if we change it here. 

Related to this, how do I know as an application what connections I affect when changing a protocol property if connections are grouped implicitly? We have in the API draft: 

> Connections in a Connection Group share all Protocol Properties that are not applicable to a Message. Changing one of these Protocol Properties on one Connection in the group changes it for all others. For example, changing "Timeout for aborting Connection" (see {{conn-timeout}}) on one Connection in a group will automatically change this Protocol Property for all Connections in the group in the same way. 

I think we would need some comment about the effect of the implicit grouping in the text above. With implicit grouping changing a protocol property may affect multiple connectione even if they were not explicitly grouped with Clone. This may not be a problem, but the semantics should be clear.

Reviewer 2: I also wanted such implicit grouping some time ago, but @abrunstrom convinced me that this doesn't add value (and indeed creates problems like she describes here). That's because "clone" only expresses a wish for Connections to be grouped in some form, but there is no promise that much is really happening under the hood. I.e., after calling "clone", an application doesn't really know if its Connections WILL be multiplexed, and if priorities between them WILL play out correctly. However, there's a consistent behavior now regarding the things that ARE promised - e.g. the Connection Properties that she describes here: if one is changed, it affects all the Connections within a Group.

As a conclusion, rather than changing the text she quotes, I think this implicit grouping should be removed. The default behavior for an application should just be to use Clone whenever possible ... whenever there's no good reason NOT to use it.

3. Both stacks MUST offer the same security properties. The inclusion of transport security protocols {{I-D.ietf-taps-transport-security}} in a Protocol Stack adds additional restrictions to Protocol Stack equivalence. Security features and properties, such as cryptographic algorithms, peer authentication, and identity privacy vary across security protocols, and across versions of security protocols. Protocol equivalence ought not to be assumed for different protocols or protocol versions, even if they offer similar application configuration options. To ensure that security protocols are not incorrectly swapped, Transport Services systems SHOULD only automatically generate equivalent Protocol Stacks when the transport security protocols within the stacks are identical. Specifically, a transport system would consider protocols identical only if they are of the same type and version. For example, the same version of TLS running over two different transport protocol stacks are considered equivalent, whereas TLS 1.2 and TLS 1.3 {{RFC8446}} are not considered equivalent.

### Separating Connection Groups {#groups}

> By default, all stored properties of the Implementation are shared within a process, such as cached protocol state, cached path state, and heuristics. This provides efficiency and convenience for the application, since the Transport System Implementation will be able to automatically optimize behavior.

Reviewer 3: nit: s/will be able to/can

Author: Fixed

Reviewer 3: 👍

> By default, all stored properties of the Implementation are shared within a process, such as cached protocol state, cached path state, and heuristics. This provides efficiency and convenience for the application, since the Transport System Implementation will be able to automatically optimize behavior.

Reviewer 4: ```suggestion
By default, all stored properties of the Implementation are shared within a process, such as cached protocol state, cached path state, and heuristics. This provides efficiency and convenience for the application, since the Transport System Implementation can automatically optimize behavior.
```

...implementing @chris-wood's suggestion as a Suggestion

There are several reasons, however, that an application might want to isolate some Connections within a single process. These reasons include:

> - Privacy concerns about re-using cached protocol state that can lead to linkability. Sensitive state may include TLS session resumption state {{RFC8446}} and HTTP cookies {{RFC6265}}.

Reviewer 3: I'd drop "resumption" here.

Author: Dropped

Reviewer 3: :+1

> - Privacy concerns about allowing Connections to multiplex together, which can tell a Remote Endpoint that all of the Connections are coming from the same application (for example, when Connections are multiplexed HTTP/2 or QUIC streams).

Reviewer 3: This is a bit of a double edged sword since multiplexing hides information from the on-path observer. I might drop this one, since coalescing to the same origin is (IMO) fine.

Reviewer 4: does the transport system implementation know enough when autocoalescing to ensure that it's the same origin, though? Thinking about cdn/cloud load balancers here...

Author: Yeah, I'd like to leave this. It's given as a concern an application may have; and when one is being paranoid (like for a private browsing case) it's nice to have a guarantee in usage like this.

- Performance concerns about Connections introducing head-of-line blocking due to multiplexing or needing to share state on a single thread.

> The Transport Services API SHOULD allow applications to explicitly define Connection Groups to force separation of Cached State and Protocol Stacks. The interface to specify these Groups can optionally expose fine-grained tuning for which properties and cached state is allowed to be shared with other Connections. For example, an application might want to allow sharing TCP Fast Open cookies across groups, but not TLS resumption state. 

Reviewer 3: Maybe add an example which says that connections for a single browser tab would be grouped?

Author: Added example!

### Separating Connection Groups {#groups}

By default, all stored properties of the Implementation are shared within a process, such as cached protocol state, cached path state, and heuristics. This provides efficiency and convenience for the application, since the Transport System Implementation can automatically optimize behavior.

There are several reasons, however, that an application might want to isolate some Connections within a single process. These reasons include:

- Privacy concerns about re-using cached protocol state that can lead to linkability. Sensitive state may include TLS session state {{RFC8446}} and HTTP cookies {{RFC6265}}.
- Privacy concerns about allowing Connections to multiplex together, which can tell a Remote Endpoint that all of the Connections are coming from the same application (for example, when Connections are multiplexed HTTP/2 or QUIC streams).
- Performance concerns about Connections introducing head-of-line blocking due to multiplexing or needing to share state on a single thread.

> The Transport Services API SHOULD allow applications to explicitly define Connection Groups to force separation of Cached State and Protocol Stacks. For example, a web browser application might use separate Connection Groups for different tabs in the browser to decrease linkability.

Reviewer 2: This would automatically be solved if you agree with my comment regarding clone above: if this is never implicitly happening but always only a result of clone, you don't even need this SHOULD here.

Author: Clarified what the SHOULD is here. The SHOULD is now that the interface should allow apps to mark their Groups as having an isolated cache. So, it says you need to be able to specify a knob to have an isolated group.

> The interface to specify these Groups can optionally expose fine-grained tuning for which properties and cached state is allowed to be shared with other Connections. For example, an application might want to allow sharing TCP Fast Open cookies across groups, but not TLS session state.

Reviewer 2: I suppose anyone can optionally add whatever bells and whistles they want to their transport system... else, I wonder, is this important?  If it is, we could e.g. add a statement to the grouping text of the interface draft to say that, for every Connection Property, a "sharing allowed" boolean could also be offered by the API... or something like that.

Author: Changed the language slightly to just say the interface MAY expose fine-grained tuning. I want to keep this in here to mitigate concerns that readers of the document may have about the architecture lumping layers together.
